src/Microsoft.ApplicationInsights.AspNetCore/DiagnosticListeners/Implementation/HostingDiagnosticListener.cs

```diff
@@ -439,18 +472,34 @@ private void AddAppIdToResponseIfRequired(HttpContext httpContext, RequestTeleme
             }
         }
 
-        private RequestTelemetry InitializeRequestTelemetry(HttpContext httpContext, Activity activity, long timestamp)
+        private static string FormatTelemetryId(string traceId, string spanId)
+        {
+            return string.Concat("|", traceId, ".", spanId, ".");
+        }
+
+        /// <summary>
+        /// Checks if the given string is a valid trace-id as per W3C Specs.
+        /// https://github.com/w3c/distributed-tracing/blob/master/trace_context/HTTP_HEADER_FORMAT.md#trace-id .
+        /// </summary>
+        /// <returns>true if valid w3c trace id, otherwise false.</returns>
+        private static bool IsCompatibleW3CTraceID(string traceId)
+        {
+            return TraceIdRegex.IsMatch(traceId);
```

Comment:
accept `Span<char>` as parameter and rather check length and each char to be in a good range than using regex.